sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/TablesJacksonSerializer.java

```diff
@@ -27,71 +25,117 @@
-                    logger.info("Node is an array of items.");
+                    // This is a multiple-entity response.
+                    for (JsonNode childEntry : childNode) {
+                        values.add(getEntityFieldsAsMap(childEntry));
+                    }
+                } else {
+                    // This is a single-entity response where the user just happened to use the key "value".
+                    if (singleValue == null) {
+                        singleValue = new HashMap<>();
+                    }
+                    singleValue.put(fieldName, getEntityFieldAsObject(node, fieldName));
                 }
-
-                throw logger.logExceptionAsError(
-                    new UnsupportedOperationException("Multiple return values not supported yet."));
+            } else {
+                // This is a single-entity response.
+                if (singleValue == null) {
+                    singleValue = new HashMap<>();
+                }
+                singleValue.put(fieldName, getEntityFieldAsObject(node, fieldName));
             }
-
-            rootObject.put(entry.getKey(), entry.getValue().asText());
         }
 
-        values.add(0, rootObject);
+        if (singleValue != null) {
+            if (values.size() > 0) {
+                throw logger.logExceptionAsError(new IllegalStateException(
+                    "Unexpected response format. Response containing a 'value' array must not contain other properties."
+                ));
+            }
+            values.add(singleValue);
+        }
 
         return (U) new TableEntityQueryResponse()
             .setOdataMetadata(odataMetadata)
             .setValue(values);
     }
 
-    @Override
-    public <U> U deserialize(HttpHeaders httpHeaders, Type type) throws IOException {
-        return super.deserialize(httpHeaders, type);
+    private Map<String, Object> getEntityFieldsAsMap(JsonNode node) {
+        Map<String, Object> result = new HashMap<>();
+        for (Iterator<String> it = node.fieldNames(); it.hasNext();) {
+            String fieldName = it.next();
+            result.put(fieldName, getEntityFieldAsObject(node, fieldName));
+
+            // If there is a field with "<name>@odata.type" = "Null",
+            // we need to insert a property with "<name>" = null.
+            if (fieldName.endsWith(TablesConstants.ODATA_TYPE_KEY_SUFFIX)
+                && TablesConstants.ODATA_TYPE_NULL.equals(node.get(fieldName).asText())) {
+
+                String nullFieldName = fieldName.substring(0, fieldName.length()
+                    - TablesConstants.ODATA_TYPE_KEY_SUFFIX.length());
+                result.put(nullFieldName, null);
+            }
+        }
+        return result;
+    }
+
+    private Object getEntityFieldAsObject(JsonNode parentNode, String fieldName) {
+        JsonNode valueNode = parentNode.get(fieldName);
+        if (!TablesConstants.METADATA_KEYS.contains(fieldName)
```

Comment:
The multi nested if loops is hard to reason about.. is it possible to flatten it and return early?
```java
if (TablesConstants.METADATA_KEYS.contains(fieldName) ....  ) {
   return valueNode.asText();
}
```

Reply:
Fixed